### PR TITLE
Enable and disable the examplers storage with flag

### LINKED ISSUE #5789: Receive: make exemplars storage a flag
### Is your proposal related to a problem?

Right now Thanos Receiver hardcoded `exemplarStorage` to true here https://github.com/thanos-io/thanos/blob/main/cmd/thanos/receive.go#L86.

For users who don't need exemplars at all, this introduces additional overhead on the TSDB side. We should make it configurable to turn on/off for this feature. Even in Prometheus this is not enabled by default.

### Describe the solution you'd like

Add a new flag in Thanos Receiver.

### PR DESCRIPTION
**Pull request by hiteshwani29**
PR closes #5789

* [x] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes

Added `--tsdb.enable-exemplar-storage` flag to enable examples storage. For enabling examples storage we should enable this `tsdb.enable-exemplar-storage` flag and set value of  `tsdb.enable-exemplar-storage` > 0. To disable examples storage set value of `tsdb.enable-exemplar-storage` <=0 and disable `tsdb.enable-exemplar-storage` flag.

## Verification

Tested locally

### CODE REVIEW

**Comment by matej-g** [comment]
Thanks for this @hiteshwani29, but wonder in line with comment https://github.com/thanos-io/thanos/issues/5789#issuecomment-1426997163, do we necessarily need a new flag?

**Comment by hiteshwani29** [comment]
Yes even if we pass `--tsdb.max-exemplars=0` it will still enables exempler storage because we are hardcoading this `EnableExemplarStorage` to `true`.

If we check this  tsdb code https://github.com/prometheus/prometheus/blob/main/tsdb/head_append.go#L56

we can figure out that it enables exempler storage` if max-exemplers is <= 0 or EnableExemplarStorage is true`
Right now this EnableExemplarStorage is always true so it enables exempler storage even if we pass max-exempler value <=0

**Comment by hiteshwani29** [comment]
Workaround you suggested will not work. With EnableExemplarStorage: false
will always disable exemplar storage. we can't able to enable or disable it
by only passing the values to the max-exemplars.
We need this flag set to enable or disable exemplar storage.

**Comment by GiedriusS** [comment]
> 

See my edited message :smile: @hiteshwani29

**Comment by hiteshwani29** [comment]
Yeah  :smile:  @GiedriusS 
Agreed to enable exemplar storage `EnableExemplarStorage: true maxExemplars > 0` will do the job
But to disable storage we need to set `EnableExemplarStorage: false`. We will not able to disable exemplars storage with `EnableExemplarStorage: true maxExemplars < 0` with current changes. We should toggle this value as per need. We should not hardcode it to `true`.

**Comment by hiteshwani29** [comment]
@GiedriusS @matej-g @yeya24 Any update on this PR?

File: cmd/thanos/receive.go [hunk 1]
```diff
@@ -85,7 +92,7 @@ func registerReceive(app *extkingpin.App) {
 			NoLockfile:                     conf.noLockFile,
 			WALCompression:                 conf.walCompression,
 			MaxExemplars:                   conf.tsdbMaxExemplars,
-			EnableExemplarStorage:          true,
+			EnableExemplarStorage:          conf.tsdbEnableExemplarStorage,
```
Comment (GiedriusS) [inline cmd/thanos/receive.go:95]: It seems like you're still missing this:
```suggestion
			EnableExemplarStorage:          conf.tsdbMaxExemplars > 0,
```
Please remove the new flag :smile:

**Comment by hiteshwani29** [comment]
Did the suggested changes @GiedriusS

File: CHANGELOG.md [hunk 2]
```diff
@@ -22,6 +22,7 @@ We use *breaking :warning:* to mark changes that are not backward compatible (re
 - [#6183](https://github.com/thanos-io/thanos/pull/6183) Receiver: fix off by one in multitsdb flush that will result in empty blocks if the head only contains one sample
 - [#6197](https://github.com/thanos-io/thanos/pull/6197) Exemplar OTel: Fix exemplar for otel to use traceId instead of spanId and sample only if trace is sampled
 - [#6207](https://github.com/thanos-io/thanos/pull/6207) Receive: Remove the shipper once a tenant has been pruned.
+- [#6216](https://github.com/thanos-io/thanos/pull/6216) Receiver: Fix Removed hardcoded value of EnableExemplarStorage flag ans set it according to max-exemplar value
```
Comment (GiedriusS) [inline CHANGELOG.md:25]:
```suggestion
- [#6216](https://github.com/thanos-io/thanos/pull/6216) Receiver: removed hardcoded value of EnableExemplarStorage flag and set it according to max-exemplar value
```

Comment (hiteshwani29) [reply CHANGELOG.md]: Done @GiedriusS

**Review by GiedriusS — APPROVED** [review]
:medal_military:

**Review by fpetkovski — APPROVED** [review]
Thanks 👍

**Pull request merged**